### ISSUE TITLE
'has no exported member' error when the component is dependencies of another component

### ISSUE THREAD

**Issue by JoshK2** [teambit/bit#1947]
### Describe the bug
I try to build this project https://github.com/ngx-fomantic-ui/ngx-fomantic-ui
and build misc component is passing, but when I build rating component that has is in dependencies the misc component I get an error that misc not exported some functions.

### Steps to Reproduce

1. clone https://github.com/ngx-fomantic-ui/ngx-fomantic-ui
2. npm i 
3. bit init
4. add all the components with this commands:
```
bit add projects/ngx-fomantic-ui/src/modules/*/ -m 'projects/ngx-fomantic-ui/src/modules/{PARENT}/public.ts'  
bit add projects/ngx-fomantic-ui/src/misc -m 'projects/ngx-fomantic-ui/src/misc/public.ts'
bit add projects/ngx-fomantic-ui/src/behaviors -m 'projects/ngx-fomantic-ui/src/behaviors/public.ts'
```
5. run bit status, need to be ok.
6. and run bit build rating to see the error.

### Screenshots, exceptions and logs
The error I get:
```
error: bit failed to build rating with the following exception:
Command failed: node /Users/yoshuakuttler/Documents/bit/testingAngularProjects/ngx-fomantic-ui-compiler/components/compilers/angular/node_modules/ng-packagr/cli/main.js -p ng-package.json -c tsconfig.json
ERROR: misc/util/internal.ts(1,10): error TS2305: Module '"../../../../../T/bit/capsule-83637/node_modules/@bit/misc/dist/misc/bit-misc"' has no exported member 'CustomValueAccessor'.
misc/util/internal.ts(3,10): error TS2305: Module '"../../../../../T/bit/capsule-83637/node_modules/@bit/misc/dist/misc/bit-misc"' has no exported member 'customValueAccessorFactory'.
misc/util/internal.ts(5,10): error TS2305: Module '"../../../../../T/bit/capsule-83637/node_modules/@bit/misc/dist/misc/bit-misc"' has no exported member 'ICustomValueAccessorHost'.
```

### Specifications

- Bit version: 14.2.5-dev.1
- Node version: 10.16.0
- npm version: 6.9.0
- Platform: macOS
- Bit compiler (include version): https://bit.dev/bit/envs/compilers/angular?version=0.1.9

**Comment by davidfirst** [comment]
@GiladShoham , this seems to be an issue with a component requires an internal file of another component, which we expected to have issues with.
In this case, the component `checkbox` requires another component `misc`. The main file of `misc` is `public_api.ts`. However, the required file is `'../../../misc/util/internal'`.

When debugging I can see how Bit doesn't find the dist of `util/internal` and as a result falls back to the main package: `import { customValueAccessorFactory } from '@bit/misc';`.
Changing the dist directly to `import { customValueAccessorFactory } from '@bit/misc/dist/misc/util/internal';` makes it work.

During the debugging, I printed all dists paths. 

```
misc/bit-misc.d.ts
misc/bit-misc.metadata.json
misc/package.json
misc/public.d.ts
misc/public_api.d.ts
misc/bundles/bit-misc.umd.js
misc/bundles/bit-misc.umd.js.map
misc/bundles/bit-misc.umd.min.js.map
misc/bundles/bit-misc.umd.min.js
misc/fesm5/bit-misc.js
misc/fesm5/bit-misc.js.map
misc/fesm2015/bit-misc.js
misc/fesm2015/bit-misc.js.map
misc/util/internal.d.ts
misc/util/public.d.ts
misc/util/util.module.d.ts
misc/util/helpers/custom-validator.d.ts
misc/util/helpers/date.d.ts
misc/util/helpers/custom-value-accessor.d.ts
misc/util/helpers/util.d.ts
misc/util/helpers/focus-event.d.ts
misc/util/services/component-factory.service.d.ts
misc/util/services/positioning.service.d.ts
misc/esm2015/bit-misc.js
misc/esm2015/public.js
misc/esm2015/public_api.js
misc/esm2015/util/internal.js
misc/esm2015/util/public.js
misc/esm2015/util/util.module.js
misc/esm2015/util/helpers/custom-validator.js
misc/esm2015/util/helpers/date.js
misc/esm2015/util/helpers/custom-value-accessor.js
misc/esm2015/util/helpers/util.js
misc/esm2015/util/helpers/focus-event.js
misc/esm2015/util/services/component-factory.service.js
misc/esm2015/util/services/positioning.service.js
misc/esm5/public.js
misc/esm5/public_api.js
misc/esm5/bit-misc.js
misc/esm5/util/internal.js
misc/esm5/util/public.js
misc/esm5/util/util.module.js
misc/esm5/util/helpers/custom-validator.js
misc/esm5/util/helpers/custom-value-accessor.js
misc/esm5/util/helpers/date.js
misc/esm5/util/helpers/focus-event.js
misc/esm5/util/helpers/util.js
misc/esm5/util/services/component-factory.service.js
misc/esm5/util/services/positioning.service.js
```
as you can see there is no `util/internal`. 

We need to find a way to solve these cases. Not going to be easy.

**Comment by davidfirst** [comment]
@JoshK2 , I discussed with Gilad, I created a new branch with his suggestion of how it could work.
It is more like a workaround, we need to figure out if it works for different cases.
Once testing is done I'll merge a permanent fix.
To work with this change, please checkout to `bug/angular-compiler-dist-prefix` branch.

**Comment by JoshK2** [comment]
@davidfirst this bug happens me also in this project: https://github.com/edcarroll/ng2-semantic-ui
and I test it with the branch `bug/angular-compiler-dist-prefix` and it works on this and on `ngx-fomantic-ui` project.

**Comment by JoshK2** [comment]
I build the rating component in `ngx-fomantic-ui` and is working, but when I build the entire project I get the following error:
```
error: bit failed to build modal with the following exception:
Command failed: node /Users/yoshuakuttler/Documents/bit/testingAngularProjects/ngx-fomantic-ui/.git/bit/components/compilers/angular/bit.envs/0.2.0/node_modules/ng-packagr/cli/main.js -p ng-package.json -c tsconfig.json
Warning: Can't resolve all parameters for FuiModalService in /private/var/folders/r8/w61z8lqd3hz5d_gmx9985tx40000gn/T/bit/capsule-38578/modules/modal/services/modal.service.ts: (?). This will become an error in Angular v6.x
ERROR: misc/util/internal.ts(1,33): error TS2307: Cannot find module '@bit/misc/dist/misc/util/internal'.
misc/util/internal.ts(3,25): error TS2307: Cannot find module '@bit/misc/dist/misc/util/internal'.
misc/util/internal.ts(5,37): error TS2307: Cannot find module '@bit/misc/dist/misc/util/internal'.
misc/util/internal.ts(7,22): error TS2307: Cannot find module '@bit/misc/dist/misc/util/internal'.
misc/util/internal.ts(9,34): error TS2307: Cannot find module '@bit/misc/dist/misc/util/internal'.


Building Angular Package

------------------------------------------------------------------------------
Building entry point '@bit/modal'
------------------------------------------------------------------------------
Compiling TypeScript sources through ngc
```

The entry point is the modal component, so I try to build it and its working when I run `bit build modal`, but when I run `bit build` to build all the components is not working.

**Comment by JoshK2** [comment]
Note that build all the project is working if I build each component in an individual build command.
Like this:
```
bd tag collapse 0.11.4 
bd tag transition 0.11.4 
bd tag accordion 0.11.4  
bd tag behaviors 0.11.4 
bd tag misc 0.11.4     
bd tag checkbox 0.11.4 
bd tag popup 0.11.4
bd tag datepicker 0.11.4 
bd tag dimmer 0.11.4  
bd tag dropdown 0.11.4 
bd tag modal 0.11.4   
bd tag progress 0.11.4 
bd tag rating 0.11.4  
bd tag search 0.11.4
bd tag select 0.11.4 
bd tag sidebar 0.11.4 
bd tag tabs 0.11.4 
bd tag toast 0.11.4
```

I used this branch: https://github.com/teambit/bit/tree/bug/angular-compiler-dist-prefix

**Comment by JoshK2** [comment]
Build both the entire projects with the new dev version is working.
Bit version: 14.2.6-dev.1

**Comment by davidfirst** [comment]
Due to limitation of NgPackgers, the component-name has to be added as a prefix for all dists path.
When Bit generates the links for dists file, there should be a way to tell the link-generator to look into the dists with the prefix of the component-name.

The initial thought was to ask the compiler to provide a dist-prefix (if it has any) and then use it for generating the link. 
This will work properly when running `bit build`. However, `bit link` won't work as it doesn't run the compiler so it doesn't know about dist-prefix. 
Our option here is to save a new property in the objects/model `distPrefix` or something similar. This will work fine in this case when all dists have the same prefix. 

An alternative would be to save a new property in the dists array in the model.
Something like "pathInSource", (in addition to what we have now `relativePath`, `file`, `name` and `test`), which saves the path of the source file corresponding to this dist file.
There won't be any backward compatibility issue as the "dists" are not part of the hash calculation.
Also, this will work for cases when the dist files are not parallel to the source. E.g. when bundling the source into one file. 
@GiladShoham , @qballer 
Please let me know your thoughts or better suggestions if you have any.

**Comment by GiladShoham** [comment]
@davidfirst I believe we already have this partially implemented.
The compiler dist vinyls suppose to have a base property.
This can be used as the dist prefix. I think it should already behave that way.
Vinyl has a computed relative property which is internally the relative between path and base.
So if @qballer will set the base correctly - like dist/compName. everything should work. 
(If we do it another way not using this relative prop somewhere it's kind of bug).
@qballer already added this base to some of the compilers.
Old compilers just set this prop to the distRoot provided by bit.
That's was the original intention of mine when moving all the writing files to use this.
(But there are a lot of new changes since I did this refactoring so might have places when we don't follow this).
Also, we might need to save this base in the models which I think we don't now.

Another related issue is that the ng-packager not only added the compName but also have a different structure with the esm2015, fesm etc'. 
So the corresponding dist is not something like `dist/compName/[relative]` but something like `dist/compName/esm2015/[relative]`.
So it might cause some issues..

**Comment by davidfirst** [comment]
Thanks Gilad.
For now, the implemented workaround is good enough. Once a better solution is needed, we'll re-open this or open a new task.
